File: JavaScript/intra.js
//- Function to calculate the variance of an array S
function variance(S) {
  const n = S.length;
  const mean = S.reduce((acc, val) => acc + val, 0) / n;
  return S.map(x => Math.pow(x - mean, 2)).reduce((acc, val) => acc + val, 0) / n;
}

//- Function to calculate the weighted variance of an array S with weights
function variance_w(S, weights) {
  const n = S.length;
  const mean = S.reduce((acc, val) => acc + val, 0) / n;
  const variance = S.map(x => Math.pow(x - mean, 2)).reduce((acc, val) => acc + val, 0) / n;
  return variance * weights.reduce((acc, val) => acc + val, 0);
}

//- Function to calculate the entropy of an array S
function entropy(S) {
  const sum = S.reduce((acc, val) => acc + val, 0);
  const probabilities = S.map(x => x / sum);
  return -probabilities.map(p => p * Math.log(p)).reduce((acc, val) => acc + val, 0);
}

//- Function to calculate the weighted entropy of an array S with weights
function entropy_w(S, weights) {
  const sum = S.reduce((acc, val) => acc + val, 0);
  const probabilities = S.map(x => x / sum);
  const entropy = -probabilities.map(p => p * Math.log(p)).reduce((acc, val) => acc + val, 0);
  return entropy * weights.reduce((acc, val) => acc + val, 0);
}

File: src/measures/intra.py
import numpy as np
import scipy.stats

#- Function to calculate the variance of an array S
def variance(S):
    return np.var(S).sum()

#- Function to calculate the weighted variance of an array S with weights
def variance_w(S, weights):
    return np.multiply(np.var(S), weights).sum()

#- Function to calculate the entropy of a n array S
def entropy(S):
    return scipy.stats.entropy(S).sum()

#- Function to calculate the weighted entropy of an array S with weights
def entropy_w(S, weights):
    return np.multiply(scipy.stats.entropy(S), weights).sum()
